File: typescript/packages/evaluators/src/threshold.ts
import { createEvaluator } from "@composable-model-graph/core";
import type { Evaluator } from "@composable-model-graph/core";

/** Options for {@link thresholdEvaluator}. */
export interface ThresholdEvaluatorOptions {
  id?: string;
  name?: string;
  /** The threshold the output is compared against. */
  threshold: number;
  /**
   * Pass condition relative to the threshold.
   * - `"atLeast"` (default): pass when `output >= threshold`.
   * - `"atMost"`: pass when `output <= threshold`.
   */
  direction?: "atLeast" | "atMost";
}

/**
 * Pass / fail an output by comparing it to a numeric threshold.
 */
export function thresholdEvaluator(
  options: ThresholdEvaluatorOptions,
): Evaluator<number> {
  const direction = options.direction ?? "atLeast";
  const { threshold } = options;

  return createEvaluator<number>({
    id: options.id ?? "threshold",
    name: options.name ?? "Threshold",
    evaluate: (output) => {
      const passed =
        direction === "atLeast" ? output >= threshold : output <= threshold;
      return {
        status: passed ? "pass" : "fail",
        score: passed ? 1 : 0,
        messages: [
          `value ${output} ${passed ? "meets" : "misses"} threshold ${threshold} (${direction})`,
        ],
        evidence: [
          { label: "value", value: output, source: "threshold" },
          { label: "threshold", value: threshold, source: "threshold" },
        ],
      };
    },
  });
}

File: python/src/composable_model_graph/evaluators/threshold.py
"""Pass / fail an output by comparing it to a numeric threshold.

(Python parity of typescript evaluators/threshold.ts.)
"""

from __future__ import annotations

from typing import Any, Literal, Optional

from ..core import Evaluator, EvaluationResult, Evidence, create_evaluator
from ..core._format import format_number as _fmt


def threshold_evaluator(
    threshold: float,
    direction: Literal["at_least", "at_most"] = "at_least",
    id: Optional[str] = None,
    name: Optional[str] = None,
) -> Evaluator:
    """Pass when `output >= threshold` ("at_least", default) or
    `output <= threshold` ("at_most")."""

    def evaluate(output: Any, target: Any, context: Any) -> EvaluationResult:
        passed = output >= threshold if direction == "at_least" else output <= threshold
        return EvaluationResult(
            status="pass" if passed else "fail",
            score=1.0 if passed else 0.0,
            messages=[
                f"value {_fmt(output)} {'meets' if passed else 'misses'} "
                f"threshold {_fmt(threshold)} ({direction})"
            ],
            evidence=[
                Evidence(label="value", value=output, source="threshold"),
                Evidence(label="threshold", value=threshold, source="threshold"),
            ],
        )

    return create_evaluator(
        id=id if id is not None else "threshold",
        name=name if name is not None else "Threshold",
        evaluate=evaluate,
    )
